src/main/scala/com/gu/subscriptions/cas/directives/ProxyDirective.scala

```diff
@@ -134,38 +138,80 @@ trait ProxyDirective extends Directives with ErrorRoute with LazyLogging {
 
-    onSuccess(validSubscription) {
-      case Some(subscription: Subscription[Paid]) =>
-        if (activation) { subscriptionService.updateActivationDate(subscription) }
-        // TODO ASAP - if deviceId and appId are provided:
-          // upsert a record in DynamoDB
-        //Since the dates are in PST, we want to make sure that we don't cut any subscription one day short
-        complete(SubscriptionExpiration(subscription.termEndDate.plusDays(1).toDateTimeAtStartOfDay(), ExpiryType.SUB))
-      case _ if subscriptionName.get.startsWith("A-S") =>
-        //no point going to CAS if this is a Zuora sub
-        notFound
-      case _ =>
-        //Go to CAS legacy /subs route
-        reject
+      if (!subsId.startsWith(emergencyTokens.prefix)) {
+        None
+      } else {
+        logger.info(s"EMERGENCY PROVIDER triggered for subscriber id:'$subsId' password: '${req.password} deviceId:${req.deviceId}")
+
+        Try(emergencyTokens.decoder.decode(subsId)) match {
+
+          case Success(payloadResult) =>
+            logger.info(s"subscriber id:'$subsId' resolves to $payloadResult")
+            processPayload(payloadResult)
+
+          case Failure(cause) =>
+            logger.error(s"error decoding token $subsId :  $cause")
+            None
+        }
+      }
     }
   }
 
+  def processSub(subsReq: SubscriptionRequest): Route = {
+    processEmergencyToken(subsReq).map(complete(_)) getOrElse {
+      zuoraDirective(subsReq) { (activation, subscriptionName) =>
+
+        val validSubscription = subscriptionService.getValidSubscription(Name(subscriptionName.get.trim.dropWhile(_ == '0')), subsReq.password.getOrElse(""))
+
+        validSubscription.onFailure {
+          case t: Throwable =>
+            logger.error(s"Failed getting Zuora subscription ${t.getMessage} ${subsReq.subscriberId}")
+            throw t
+        }
+
+        onSuccess(validSubscription) {
+          case Some(subscription: Subscription[Paid]) =>
+            if (activation) {
+              subscriptionService.updateActivationDate(subscription)
+            }
+            // TODO ASAP - if deviceId and appId are provided:
+            // upsert a record in DynamoDB
+            //Since the dates are in PST, we want to make sure that we don't cut any subscription one day short
+
+            complete(SubscriptionExpiration(subscription.termEndDate.plusDays(1).toDateTimeAtStartOfDay(), ExpiryType.SUB))
+          case _ => notFound
+        }
+      }
+    }
+  }
   val subsRoute = (path("subs") & post) {
     entity(as[SubscriptionRequest]) { subsReq =>
       // TODO third - handle limit of regisrations
         // get count of activations for these credentials from Dynamo
         // if count >= "max.subscriptions.per.user" return error: "Credentials used too often", credentials.overuse.error.code
         // else, continue, the zuoraRoute must update the count iff successful
-      zuoraRoute(subsReq) ~ casRoute
+      processSub(subsReq)
```

Comment:
not keen on the 'process' terminology as it is too generic, can we call this something more specific?